forge-core/src/main/java/forge/token/TokenDb.java

```diff
@@ -38,38 +43,79 @@ public boolean containsRule(String rule) {
-        return getToken(tokenName, CardEdition.UNKNOWN.getName());
-    }
 
     public void preloadTokens() {
         for (CardEdition edition : this.editions) {
-            for (String name : edition.getTokens().keySet()) {
-                try {
-                    getToken(name, edition.getCode());
-                } catch(Exception e) {
-                    System.out.println(name + "_" + edition.getCode() + " defined in Edition file, but not defined as a token script.");
+            for (Map.Entry<String, Collection<CardEdition.TokenInSet>> inSet : edition.getTokens().asMap().entrySet()) {
+                String name = inSet.getKey();
+                String fullName = String.format("%s_%s", name, edition.getCode().toLowerCase());
+                for (CardEdition.TokenInSet t : inSet.getValue()) {
+                    allTokenByName.put(fullName, addTokenInSet(edition, name, t));
                 }
             }
         }
     }
 
+    protected boolean loadTokenFromSet(CardEdition edition, String name) {
+        String fullName = String.format("%s_%s", name, edition.getCode().toLowerCase());
+        if (allTokenByName.containsKey(fullName)) {
+            return true;
+        }
+        if (!edition.getTokens().containsKey(name)) {
+            return false;
+        }
+
+        for (CardEdition.TokenInSet t : edition.getTokens().get(name)) {
+            allTokenByName.put(fullName, addTokenInSet(edition, name, t));
+        }
+        return true;
+    }
+
+    protected PaperToken addTokenInSet(CardEdition edition, String name, CardEdition.TokenInSet t) {
+        return new PaperToken(rulesByName.get(name), edition, name, t.collectorNumber, t.artistName);
+    }
+
+    // try all editions to find token
+    protected PaperToken fallbackToken(String name) {
+        for (CardEdition edition : this.editions) {
+            String fullName = String.format("%s_%s", name, edition.getCode().toLowerCase());
+            if (loadTokenFromSet(edition, name)) {
+                return Aggregates.random(allTokenByName.get(fullName));
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public PaperToken getToken(String tokenName) {
+        return getToken(tokenName, CardEdition.UNKNOWN.getCode());
+    }
+
     @Override
     public PaperToken getToken(String tokenName, String edition) {
-        String fullName = String.format("%s_%s", tokenName, edition.toLowerCase());
+        CardEdition realEdition = editions.getEditionByCodeOrThrow(edition);
+        String fullName = String.format("%s_%s", tokenName, realEdition.getCode().toLowerCase());
+
+        // token exist in Set, return one at random
+        if (loadTokenFromSet(realEdition, tokenName)) {
+            return Aggregates.random(allTokenByName.get(fullName));
+        }
+        PaperToken fallback = this.fallbackToken(tokenName);
+        if (fallback != null) {
+            return fallback;
+        }
```

Comment:
@tehdiplomat @tool4ever i added a fallback in case the Token isn't found in the right set so it tries any other set

i don't know yet if that makes the last part with "extraTokenByName" obsolete